libs/langchain-community/src/chat_models/moonshot.ts

```diff
@@ -0,0 +1,470 @@
+   */
+  temperature?: number;
+
+  /** Total probability mass of tokens to consider at each step. Range
+   * from 0 to 1 Defaults to 0.7
+   */
+  topP?: number;
+
+  /**
+   * max value is 8192，defaults to 1024
+   */
+  maxTokens?: number;
+
+  stop?: string[];
+
+  /**
+   * There is a penalty, a number between -2.0 and 2.0. Positive values
+   * ​​penalize the newly generated words based on whether they appear in the
+   * text, increasing the likelihood that the model will discuss new topics.
+   * The default value is 0
+   */
+  presencePenalty?: number;
+
+  /**
+   * Frequency penalty, a number between -2.0 and 2.0. Positive values
+   * ​penalize the newly generated words based on their existing frequency in the
+   * text, making the model less likely to repeat the same words verbatim.
+   * The default value is 0
+   */
+  frequencyPenalty?: number;
+
+  /**
+   * The default value is 1 and cannot be greater than 5. In particular,
+   * when temperature is very small and close to 0, we can only return 1 result.
+   * If n is already set and > 1, Moonshot will return an invalid input parameter
+   * (invalid_request_error).
+   */
+  n?: number;
+}
+
+function messageToRole(message: BaseMessage): MoonshotMessageRole {
+  const type = message._getType();
+  switch (type) {
+    case "ai":
+      return "assistant";
+    case "human":
+      return "user";
+    case "system":
+      return "system";
+    case "function":
+      throw new Error("Function messages not supported yet");
+    case "generic": {
+      if (!ChatMessage.isInstance(message)) {
+        throw new Error("Invalid generic chat message");
+      }
+      if (["system", "assistant", "user"].includes(message.role)) {
+        return message.role as MoonshotMessageRole;
+      }
+      throw new Error(`Unknown message type: ${type}`);
+    }
+    default:
+      throw new Error(`Unknown message type: ${type}`);
+  }
+}
+
+export class ChatMoonshot extends BaseChatModel implements ChatMoonshotParams {
+  static lc_name() {
+    return "ChatMoonshot";
+  }
+
+  get callKeys() {
+    return ["stop", "signal", "options"];
+  }
+
+  get lc_secrets() {
+    return {
+      moonshotApiKey: "Moonshot_API_KEY",
+      apiKey: "Moonshot_API_KEY",
+    };
+  }
+
+  get lc_aliases() {
+    return undefined;
+  }
+
+  moonshotApiKey?: string;
+
+  apiKey?: string;
+
+  streaming: boolean;
+
+  messages?: MoonshotMessage[];
+
+  modelName: ChatCompletionRequest["model"];
+
+  model: ChatCompletionRequest["model"];
+
+  apiUrl: string;
+
+  maxTokens?: number | undefined;
+
+  temperature?: number | undefined;
+
+  topP?: number | undefined;
+
+  stop?: string[];
+
+  presencePenalty?: number;
+
+  frequencyPenalty?: number;
+
+  n?: number;
+
+  constructor(fields: Partial<ChatMoonshotParams> & BaseChatModelParams = {}) {
+    super(fields);
+
+    this.moonshotApiKey =
+      fields?.apiKey ??
+      fields?.moonshotApiKey ??
+      getEnvironmentVariable("Moonshot_API_KEY");
```

Comment:
Should be all caps

Reply:
Yes, environment variables need to be capitalized. Already updated.